File: devenv/soul.parsing/Action.cs
/*========================================================================
    Copyright (c) 2011-2012 Seppo Laakko
    http://sourceforge.net/projects/soulparsing/
 
    Distributed under the GNU General Public License, version 3 (GPLv3).
    (See accompanying LICENSE.txt or http://www.gnu.org/licenses/gpl.html)

========================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using soul.codedom;

namespace soul.parsing
{
    public delegate void ParsingAction(string match, string content, Position position, string fileName, ref bool pass);
    public delegate void FailureAction();

    public class ActionParser : UnaryParser, INamedEntity
    {
        public ActionParser(string name, Parser child)
            : this(name, null, child)
        {
        }
        public ActionParser(string name, BlockStatement sucessCode, Parser child)
            : base(child, child.Info)
        {
            this.name = name;
            this.successCode = sucessCode;
        }
        public ActionParser(string name, BlockStatement successCode, BlockStatement failCode, Parser child)
            : base(child, child.Info)
        {
            this.name = name;
            this.successCode = successCode;
            this.failCode = failCode;
        }
        public string Name
        {
            get { return name; }
            set { name = value; }
        }
        public string FullName
        {
            get { return name; }
        }
        public string MethodName
        {
            get { return name + "Action"; }
        }
        public string VariableName
        {
            get { return (name.Length > 0 ? char.ToLower(name[0]) + name.Substring(1) : name) + "ActionParser"; }
        }
        public BlockStatement SuccessCode
        {
            get { return successCode; }
        }
        public BlockStatement FailCode
        {
            get { return failCode; }
        }
        public ParsingAction Action
        {
            get { return action; }
            set { action = value; }
        }
        public FailureAction FailureAction
        {
            get { return failureAction; }
            set { failureAction = value; }
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position position = scanner.Position;
            Match match = Child.Parse(scanner, stack);
            if (match.Hit)
            {
                if (action != null)
                {
                    bool pass = true;
                    action(scanner.Match(position.Index), scanner.Content, position, scanner.FileName, ref pass);
                    if (!pass)
                    {
                        return Match.Nothing();
                    }
                }
            }
            else if (failureAction != null)
            {
                failureAction();
            }
            return match;
        }
        public override void Accept(Visitor visitor)
        {
            visitor.BeginVisit(this);
            Child.Accept(visitor);
            visitor.EndVisit(this);
        }
        private string name;
        private BlockStatement successCode;
        private BlockStatement failCode;
        private ParsingAction action;
        private FailureAction failureAction;
    }
}

File: cmajor/parsing/Action.cpp
// =================================
// Copyright (c) 2019 Seppo Laakko
// Distributed under the MIT license
// =================================

#include <cmajor/parsing/Action.hpp>
#include <cmajor/parsing/Visitor.hpp>
#include <cctype>

namespace cmajor { namespace parsing {

ParsingAction::~ParsingAction()
{
}

FailureAction::~FailureAction()
{
}

PreCall::~PreCall()
{
}

PostCall::~PostCall()
{
}

ActionParser::ActionParser(const std::u32string& name_, Parser* child_):
    UnaryParser(name_, child_, child_->Info()), successCode(), failCode(), action(), failureAction() 
{
}

ActionParser::ActionParser(const std::u32string& name_, cmajor::codedom::CompoundStatement* successCode_, Parser* child_): 
    UnaryParser(name_, child_, child_->Info()), successCode(successCode_), failCode(), action(), failureAction() 
{
}

ActionParser::ActionParser(const std::u32string& name_, cmajor::codedom::CompoundStatement* successCode_, cmajor::codedom::CompoundStatement* failCode_, 
    Parser* child_): UnaryParser(name_, child_, child_->Info()), successCode(successCode_), failCode(failCode_), action(), failureAction() 
{
}

std::u32string ActionParser::MethodName() const 
{ 
    return Name() + U"Action"; 
}

std::u32string ActionParser::VariableName() const 
{ 
    return (Name().length() > 0 ? std::u32string(1, std::tolower(Name()[0])) + Name().substr(1) : Name()) + U"ActionParser"; 
}

Match ActionParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    Span actionSpan = scanner.GetSpan();
    Match match = Child()->Parse(scanner, stack, parsingData);
    if (match.Hit())
    {
        if (action)
        {
            bool pass = true;
            actionSpan.SetEnd(scanner.GetSpan().Start());
            const char32_t* matchBegin = scanner.Start() + actionSpan.Start();
            const char32_t* matchEnd = scanner.Start() + actionSpan.End();
            (*action)(matchBegin, matchEnd, actionSpan, scanner.FileName(), parsingData, pass);
            if (!pass)
            {
                return Match::Nothing();
            }
        }
    }
    else if (failureAction)
    {
        (*failureAction)(parsingData);
    }
    return match;
}

void ActionParser::Accept(Visitor& visitor)
{
    visitor.BeginVisit(*this);
    Child()->Accept(visitor);
    visitor.EndVisit(*this);
}

} } // namespace cmajor::parsing
